sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,171 @@
+import io.sentry.ProfilingTraceData;
+import io.sentry.SentryLevel;
+import io.sentry.util.Objects;
+import java.io.File;
+import java.util.UUID;
+import java.util.concurrent.Future;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class AndroidTraceTransactionListener implements ITransactionListener {
+
+  /**
+   * This appears to correspond to the buffer size of the data part of the file, excluding the key
+   * part. Once the buffer is full, new records are ignored, but the resulting trace file will be
+   * valid.
+   *
+   * <p>30 second traces can require a buffer of a few MB. 8MB is the default buffer size for
+   * [Debug.startMethodTracingSampling], but 3 should be enough for most cases. We can adjust this
+   * in the future if we notice that traces are being truncated in some applications.
+   */
+  private static final int BUFFER_SIZE_BYTES = 3_000_000;
+
+  private int intervalUs;
+  private @Nullable File traceFile = null;
+  private @Nullable File traceFilesDir = null;
+  private @Nullable Future<?> scheduledFinish = null;
+  private volatile @Nullable ITransaction activeTransaction = null;
+  private volatile @Nullable ProfilingTraceData lastProfilingData = null;
+  private final @NotNull SentryAndroidOptions options;
+
+  public AndroidTraceTransactionListener(final @NotNull SentryAndroidOptions sentryAndroidOptions) {
+    this.options = Objects.requireNonNull(sentryAndroidOptions, "SentryAndroidOptions is required");
+    final String tracesFilesDirPath = options.getProfilingTracesDirPath();
+    if (tracesFilesDirPath == null || tracesFilesDirPath.isEmpty()) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.WARNING,
+              "Disabling profiling because no profiling traces dir path is defined in options.");
+      return;
+    }
+    long intervalMillis = options.getProfilingTracesIntervalMillis();
+    if (intervalMillis <= 0) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.WARNING,
+              "Disabling profiling because trace interval is set to %d milliseconds",
+              intervalMillis);
+      return;
+    }
+    intervalUs = (int) MILLISECONDS.toMicros(intervalMillis);
+    traceFilesDir = new File(tracesFilesDirPath);
+  }
+
+  @Override
+  public synchronized void onTransactionStart(@NotNull ITransaction transaction) {
+
+    // Debug.startMethodTracingSampling() is only available since Lollipop
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return;
+
+    // traceFilesDir is null or intervalUs is 0 only if there was a problem in the constructor, but
+    // we already logged that
+    if (traceFilesDir == null || intervalUs == 0) {
+      return;
+    }
+
+    // If a transaction is currently being profiled, we ignore this call
+    if (activeTransaction != null) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.WARNING,
```

Comment:
@bruno-garcia suggested `WARNING`, as the user may wonder why a concurrent transaction is not profiled, and since it's an edge case having a warning looked fine